### PR TITLE
35 method nonspatial scanpy

### LINKED ISSUE #35: method_nonSpatial_ScanPy

### PR DESCRIPTION
**Pull request by theinvisibleliya**
There is still an issue (#139), which can be added later into scanpy. Scanpy uses the resolution parameter to define the number of clusters, but we require num_clusters for the future metrics. There exists a working function for an extensive search of the resolution parameter for the right number of clusters, but there are licensing issues (more in issue #139). Right now, my script throws a warning about it and the resolution parameter can be defined outside by the config.json file. One sample config_1.json is provided.

### CODE REVIEW

**Review by niklasmueboe — CHANGES_REQUESTED** [review]
Please delete all the unnecessary files first!

**Comment by theinvisibleliya** [comment]
Yep, sorry, my bad

File: .gitignore
Comment (niklasmueboe) [inline .gitignore:1]: Please remove gitignore file

File: method/scanpy/method_scanpy.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    # throw warning about not using the num_clusters as a parameter, because scanpy uses leiden or louvain and needs the resolution parameter, which is defined in the config.json. There exists a good function to perform an extensive search for the right resolution parameter to define the desired num_clusters, but we still have licensing issues. For more, see the SpaceHack2.0 GitHub issue #139
+import warnings
+
+warnings.warn("Scanpy uses leiden/louvain for clustering, which relies on the resolution parameter in the config file. The parameter num_clusters will be ignored.", UserWarning)
+
+# scanpy starts here
+sc.pp.normalize_total(adata)
+sc.pp.log1p(adata)
+sc.tl.pca(adata)
+sc.pp.neighbors(adata, n_neighbors = config["n_neighbors"], n_pcs=config["n_pcs"], random_state=seed)
```
Comment (niklasmueboe) [inline method/scanpy/method_scanpy.py:161]: If you use the dim_red file as input you don't need to preprocess the counts nor run PCA.

File: method/scanpy/scanpy_env.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+name: scanpy_env
```
Comment (niklasmueboe) [inline method/scanpy/scanpy_env.yaml:1]: Please remove the env name. Not needed

File: method/scanpy/scanpy_env.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  - defaults
+dependencies:
+  - python>=3.10
+  - numpy
+  - pandas
+  - scipy
+  - pillow
+  - anndata
+  - leidenalg
+  - scanpy
```
Comment (niklasmueboe) [inline method/scanpy/scanpy_env.yaml:13]: add major.minor.path version to all tools

File: method/scanpy/scanpy_env.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+name: scanpy_env
+channels:
+  - conda-forge
+  - defaults
```
Comment (niklasmueboe) [inline method/scanpy/scanpy_env.yaml:4]: defaults not needed

**Review by niklasmueboe — CHANGES_REQUESTED** [review]

File: method/scanpy/scanpy_env.yaml
Comment (niklasmueboe) [inline method/scanpy/scanpy_env.yaml:1]: I just menat remove the defaults channel not remove all channels.

**Review by niklasmueboe — CHANGES_REQUESTED** [review]

File: method/scanpy/method_scanpy.py [hunk 5]
```diff
@@ -90,36 +90,44 @@
...
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
 
-    if args.image is not None:
-        adata.uns["image"] = np.array(Image.open(args.image))
+    if args.img is not None:
+        adata.uns["image"] = np.array(Image.open(args.img))
```
Comment (niklasmueboe) [inline method/scanpy/method_scanpy.py:130]: Why this change?

File: method/scanpy/scanpy_env.yaml
Comment (theinvisibleliya) [reply method/scanpy/scanpy_env.yaml:1]: > I just menat remove the defaults channel not remove all channels.

yeah, that was not very intentional:( thank you!

File: method/scanpy/method_scanpy.py (hunk 5 above)
Comment (theinvisibleliya) [reply method/scanpy/method_scanpy.py:130]: > Why this change?

This is the way it is written in the original template "method.py" in the SpaceHack2. For some reason, I had "image" written out, but in your template it is "img"

Comment (niklasmueboe) [reply method/scanpy/method_scanpy.py:130]: Four days and you are the first to notice that the template is wrong! It actually has to be `args.image` if you look at the CLI. Can you change it back. And sorry for the wrong template

**Review by niklasmueboe — APPROVED** [review]

**Review by niklasmueboe — APPROVED** [review]

File: method/scanpy/method_scanpy.py (hunk 5 above)
Comment (theinvisibleliya) [reply method/scanpy/method_scanpy.py]: > Four days and you are the first to notice that the template is wrong! It actually has to be `args.image` if you look at the CLI. Can you change it back. And sorry for the wrong template

Haha nice! No worries at all, I am glad we sorted it out:)
